### ISSUE TITLE
PWA seems break ?

### ISSUE THREAD

**Issue by OnlyFor** [photoview/photoview#938]
**Describe the bug**
 
PWA is not working

**To Reproduce**
1. open chrome
2. open console
3. goto https://photos.qpqp.dk/login (or self lastest build)

**Expected behavior**

install pwa show and can install

**Screenshots**

in network console shows:

"index.f9413d49.js:224 Error during service worker registration: DOMException: Failed to register a ServiceWorker for scope ('https://photos.qpqp.dk/') with script ('https://photos.qpqp.dk/service-worker.js'): The script has an unsupported MIME type ('text/html')."

**Your environment:**
Feel free to remove and add what you think might be useful.

 - Are you using Docker? [yes / no] no 
 - Database [e.g. mariadb 10.5] sqlite
 - Browser [e.g. chrome, safari]  chrome
 - Version [e.g. 2.0.0] office site same as lastest build

**Additional context**

**Comment by kkovaletp** [comment]
@OnlyFor, I checked the image `viktorstrate/photoview:master` and don't see this error. The demo site is running the latest released image, which is pretty old. Can you please test the PWA installation from your side using the master tag for the image?

**Comment by kkovaletp** [comment]
Assuming that you used the source code from `master` (hope that pulled not too long time ago), it should work.

@jordy2254, can you please take a look, as I never built the project locally?

**Comment by jordy2254** [comment]
I believe this is the way that we serve files through go where it doesn't add the correct type onto the file when serving it thus the browser blocks it. It should be a straight forward fix, and the association for the serve handling in the go code can be found here; https://github.com/photoview/photoview/blob/master/api/server.go#L98
@OnlyFor  Shouldn't be too complicated of a fix if you have the knowhow to controbute that would be great! If not thanks for raising the contribution and just raising bugs and helping identify replication steps is a great help!

**Comment by h4de5** [comment]
I do not think this is about file associations. the service-worker file is not there and its just the 404 error page that has the wrong mime-type.

**Comment by poisonnuke** [comment]
I have the same issue with the latest docker installation. Fresh Debian server with nginx as reverse-proxy and standard-configuration. 

The serviceworker.js file cannot be found. 

looks like the solution is here: 
https://stackoverflow.com/questions/49566059/service-worker-registration-error-unsupported-mime-type-text-html

**Comment by kkovaletp** [comment]
> The serviceworker.js file cannot be found.

its strange, but I cannot reproduce the issue. I've tried using the latest `master` image from the server in Chrome and Firefox, then from a remote host in the same browsers: in all cases I see in the Network tab of Dev Tools that the `service-worker.js` is loaded with 200 OK status.

I found the next line in the code:
`const swUrl = '${import.meta.env.BASE_URL}service-worker.js'`

If you have this issue in your setup, please make sure that the `PHOTOVIEW_LISTEN_IP` variable is set to the hostname of the photoview service in your docker-compose file. In my case it is the `PHOTOVIEW_LISTEN_IP: "photoview"`

**Comment by kkovaletp** [comment]
> can you post the first few lines of content of that service-worker.js file from the Browsers dev tools?

```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <link rel="icon" href="/photoview-logo.svg" type="image/svg+xml" />
    <link rel="apple-touch-icon" href="/logo192.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="theme-color" content="#000000" />

    <link rel="manifest" href="/manifest.json" />

    <meta name="apple-mobile-web-app-title" content="Photoview" />
    <meta name="apple-mobile-web-app-capable" content="yes" />
    <meta name="apple-mobile-web-app-status-bar-style" content="white" />

    <script type="module" crossorigin src="/assets/index.9b57b0a0.js"></script>
    <link rel="stylesheet" href="/assets/index.0ac106ed.css">
  </head>
  <body>
    <noscript>You need to enable JavaScript to run Photoview.</noscript>
    <div id="root"></div>
    
  </body>
</html>
```

**Comment by poisonnuke** [comment]
In my case the service-worker.js returns a 404 page. So I cannot share any code.

The LISTEN_IP is still unchanged and set to "photoview".
Ive tried different settings (just the subdomain of my host, or the entire domain (sub.domain.tld), but that didnt change anything. What is that value supposed to be?

**Comment by kkovaletp** [comment]
> The LISTEN_IP is still unchanged and set to "photoview".

it depends on your deployment type:
- if you use docker, then it should be set to the Photoview service's hostname in the docker-compose file. In most cases that is "photoview"
- if you install Photoview on the host directly, then this should be the environment variable, set to the host IP or hostname. Try local network IP or name and test from the local network: if the issue is gone, then this is the cause and you need to find a way to configure it to work for all use cases

**Comment by poisonnuke** [comment]
it runs with docker compose. Therefore it was already correct with "photoview".
but if this is used to determine the URL for the service-worker shouldnt it be set to the actual domain thats being used even if everything runs inside Docker?

**Comment by kkovaletp** [comment]
> shouldn't it be set to the actual domain

in my case it works fine with the "photoview" value in the compose file, but feel free to try)

then, I don't have any other thoughts as to why it doesn't work for you((

**Comment by h4de5** [comment]
> > can you post the first few lines of content of that service-worker.js file from the Browsers dev tools?
> 
> ```
> <!DOCTYPE html>
> <html lang="en">
>   <head>
>     <meta charset="utf-8" />
>     <link rel="icon" href="/photoview-logo.svg" type="image/svg+xml" />
>     <link rel="apple-touch-icon" href="/logo192.png" />
>     <meta name="viewport" content="width=device-width, initial-scale=1" />
>     <meta name="theme-color" content="#000000" />
> 
>     <link rel="manifest" href="/manifest.json" />
> 
>     <meta name="apple-mobile-web-app-title" content="Photoview" />
>     <meta name="apple-mobile-web-app-capable" content="yes" />
>     <meta name="apple-mobile-web-app-status-bar-style" content="white" />
> 
>     <script type="module" crossorigin src="/assets/index.9b57b0a0.js"></script>
>     <link rel="stylesheet" href="/assets/index.0ac106ed.css">
>   </head>
>   <body>
>     <noscript>You need to enable JavaScript to run Photoview.</noscript>
>     <div id="root"></div>
>     
>   </body>
> </html>
> ```

you see? this is not the service worker which should be a javascript file, but instead is just the file-not page that comes with status code 200.

the service worker seems not to bundled correctly and is therefore not in the production build and the deployment (at least not under that url).

**Comment by kkovaletp** [comment]
I played with this issue a bit today and could not find any solution:
- BASE_URL is set to `/` at image build time. At the container execution time, it is unset.
- `npm run build -- --base=/` doesn't even try to build the `serviceWorkerRegistration.ts` even if it is imported in the `index.tsx`

So, we need some help from a frontend dev with experience in docker.

**Comment by kkovaletp** [comment]
> /assets directory does not exist !
> I removed the /assets and PWA worked fine.

@QuentinCazelles, thanks for your finding. I think that the issue must be related to this, however, it is not so easy:
- in the source code there are 2 manifests: `ui/src/manifest.webmanifest` and `ui/public/manifest.json`
- I'd assume that the `ui/src/manifest.webmanifest` is used at the build time, while the `ui/public/manifest.json` is used by the app when it runs.
- in the source code the `assets` folder doesn't exist, but instead the `public` folder exists with the content of the assets folder
- inside the Docker image there is the `/app/ui/assets` folder, but it doesn't contain files from the source code `ui/public` folder

So, I think that we shouldn't remove the `assets` from the manifest. instead, we need to find out why the files from the `public` folder are not copied into the `assets` folder at the build time.

**Comment by kkovaletp** [comment]
I played a bit today with paths in manifest files:
- it looks correct to remove the `/assets` from the icons' paths in the `ui/public/manifest.json`. Also, if I'm not wrong, the same paths in the `ui/src/manifest.webmanifest` should be updated by replacing `assets` with `public`.
- However, nothing from these modifications (as well as the complete removal of the icons from both manifests) didn't make `service-worker.js` to be generated in the `assets` folder in the resulting docker image, which means that it is not sent to the clients when requested.

So, while these modifications look correct, they don't make `service-worker.js` appear.

**Comment by kkovaletp** [comment]
Today I debugged this case a bit and found a possible reason why the `service-worker.js` is not generated:
It looks like we need to add the `vite-plugin-pwa` to the `vite.config.ts`, which requires the `vite` version to be upgraded to at least 3.1.0. 
Unfortunately, this is a breaking change and I don't know how to manage that, so we still need a Frontend developer for this fix.

**Comment by kkovaletp** [comment]
When working on a fix for this issue, please also check the #574, as it might be located in the same code

**Comment by kkovaletp** [comment]
Just to note: this issue was fixed in my forked repo by merging the same fix, which is waiting in #1165

It can be used as an external UI together with this image serving API: https://github.com/kkovaletp/photoview/blob/master/ui/Separate-UI-README.md
